tables: fix multiple `OrderedTable` bugs

zerbina opened the pull request:
## Summary
The fixed issues:
- for `OrderedTable`s with a single item, the iterators would get stuck in an infinite loop in some cases
- items would sometimes not be yielded when using any of the iterators
- comparing two `OrderedTable` would sometimes ignore items

These only happened for `OrdererdTable` not explicitly initialized via `initOrderedTable` - explicitly initialized ones were not affected.

## The problem
The underlying issue was the same in each case. An `OrderedTable` uses an intrusive linked list internally. Each slot stores a `next` pointer and the table itself the head (`first`) and tail (`last`) pointer.

Since the pointers are indices into the internal data `seq` and not memory addresses, the default value (i.e. 0) does not represent "none" (because the 0-th `seq` item is a valid item), meaning that the pointers have to be explicitly initialized to the "none" state (i.e. `-1`). This happens when using `initOrderedTable`, and at one point did also happen when inserting into a default-initialized table, but the latter behaviour got removed as part of
#161 (likely by accident).

When inserting an item into a default initialized `OrderedTable` and the selected slot is slot `0`, the slot's `next` pointer was set to point to the slot itself, due to the `last` pointer incorrectly pointing to slot `0` at that point. If any of the iterators were used directly after this insertion, they would loop forever (because of the self-cycle).

When inserting an item into a default initialized `OrderedTable` and the selected slot is not `0`, the `next` pointer of the unfilled slot `0` (which is treated as the head of the linked list) is set to the index of the newly filled slot. As most of the `OrderedTable` implementation checks for and skips unfilled slots, this doesn't cause any immediate problems, but once the slot at position `0` becomes used (before the table is first enlarged), all previously inserted items are effectively cut off. They're still part of the table, but are no longer returned by any of the iterators nor considered during comparisons or when using `sort`.

## Details
- add an overload of the `initImpl` template that correctly sets the `first` and `last` pointer to `-1` as part when initializing the `OrderedTable` (either explicitly or lazily)
- remove the `isFilled` guards from logic that traverses the linked list - they're unnecessary
- remove the `rawGetKnownHC`, `rawGetDeep`, and `rawGet` overloads for `OrderedTable`. They were either unused or the same as the more general implementation

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

The issue caused Nim-CPS to sometimes get stuck in an infinite loop at compile-time.

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented on lib/pure/collections/tables.nim:
```diff
@@ -1206,18 +1206,17 @@ type
-proc rawGet[A, B](t: OrderedTable[A, B], key: A, hc: var Hash): int =
-  rawGetImpl()
+template initImpl(t: OrderedTable, size: int) =
+  ## Initializes `t` to a valid empty state. Called internally by the
+  ## ``checkIfInitialized`` to set lazily set up the table before it is used
```
```suggestion
  ## Initializes `t` to a valid empty state. Called internally by
  ## ``checkIfInitialized`` to lazily set up the table before it is used
```

zerbina replied:
Thanks for reporting these. I tend to adjust documentation comments multiple times, and sometimes the remnants of previous iterations get past my proof-reading.

saem commented on lib/pure/collections/tables.nim:
```diff
@@ -1232,22 +1231,23 @@ proc enlarge[A, B](t: var OrderedTable[A, B]) =
+    var j: Hash = eh and maxHash(t)
+    # search for a free slot:
+    while isFilled(t.data[j].hcode):
+      j = nextTry(j, maxHash(t))
```
dumb question, but it doesn't look like we're mutating `t` in the loop, so do we need to continually recalculate `maxHash(t)`?

(I'm super unfamiliar with this code, again, sorry)

zerbina replied:
Your reading is correct, `maxHash` can be precomputed - the efficiency improvement is *likely* going to be minimal however, as `maxHash` is a template that expands to `t.data.high`. It is an improvement nonetheless, so I changed it.

saem commented on tests/stdlib/types/collections/torderedtables_regression.nim:
```diff
@@ -0,0 +1,57 @@
+discard """
+  description: "A regression test for lazily initialized `OrderedTable`s"
+  target: native
```
```suggestion
  target: native
  timeout: "10"
```

We're testing an infinite loop, I figure an explicit timout make sense? 10 might be too high.

zerbina replied:
Only a single iteration is expected, so an assertion is raised if a second one happens. In other words, the infinite loop is prevented and no intervention from testament's side is required.

saem replied:
Got it, thanks.

saem reviewed (approved):
A few small things, nothing is a major blocker.

zerbina committed:
tables: fix multiple `OrderedTable` bugs

## Summary
The fixed issues:
- for `OrderedTable`s with a single item, the iterators would get stuck
  in an infinite loop in some cases
- items would sometimes not be yielded when using any of the iterators
- comparing two `OrderedTable` would sometimes ignore items

These only happened for `OrdererdTable` not explicitly initialized via
`initOrderedTable` - explicitly initialized ones were not affected.

## The problem
The underlying issue was the same in each case. An `OrderedTable` uses
an intrusive linked list internally. Each slot stores a `next` pointer
and the table itself the head (`first`) and tail (`last`) pointer.

Since the pointers are indices into the internal data `seq` and not
memory adresses, the default value (i.e. 0) does not represent "none"
(because the 0-th `seq` item is a valid item), meaning that the
pointers have to be explicitly initialized to the "none" state (i.e.
`-1`). This happens when using `initOrderedTable`, and at one point did
also happen when inserting into a default-initialized table, but the
latter behaviour got removed as part of
#22e84d869efae228dc2859a9fe1fe303b6bd58f7 (likely by accident).

When inserting an item into a default initialized `OrderedTable` and
the selected slot is slot `0`, the slot's `next` pointer was set to
point to the slot itself, due to the `last` pointer incorrectly
pointing to slot `0` at that point. If any of the iterators were used
directly after this insertion, they would loop forever (because of the
self-cycle).

When inserting an item into a default initialized `OrderedTable` and the
selected slot is not `0`, the `next` pointer of the unfilled slot `0`
(which is treated as the head of the linked list) is set to the index
of the newly filled slot. As most of the `OrderedTable` implementation
checks for and skips unfilled slots, this doesn't cause any immediate
problems, but once the slot at position `0` becomes used (before the
table is first enlarged), all previously inserted items are effectively
cut off. They're still part of the table, but are no longer returned
by any of the iterators nor considered during comparisons or when using
`sort`.

## Details
- add an overload of the `initImpl` template that correctly sets the
  `first` and `last` pointer to `-1` as part when initializing the
  `OrderedTable` (either explicitly or lazily)
- remove the `isFilled` guards from logic that traverses the linked
  list - they're unnecessary
- remove the `rawGetKnownHC`, `rawGetDeep`, and `rawGet` overloads for
  `OrderedTable`. They were either unused or the same as the more
  general implementation

zerbina commented:
bors r+

saem commented:
Bors r+

I didn't take a close look, but just eliminating random failure.
